src/PreferencesController.ts

```diff
@@ -94,11 +132,67 @@ export class PreferencesController extends BaseController<BaseConfig, Preference
 		return newTokens;
 	}
 
+	/**
+	 * Adds a collectible to the stored collectible list
+	 *
+	 * @param address - Hex address of the collectible contract
+	 * @param tokenId - The NFT identifier
+	 * @returns - Current collectible list
+	 */
+	async addCollectible(address: string, tokenId: number) {
+		address = toChecksumAddress(address);
+		const existingEntry = this.state.collectibles.filter(
+			(collectible) => collectible.address === address && collectible.tokenId === tokenId
+		);
+		if (existingEntry.length > 0) {
+			return;
+		}
+		const { name, image } = await this.requestNFTCustomInformation(address, tokenId);
+		const newEntry: Collectible = { address, tokenId, name, image };
+		const collectibles = this.state.collectibles;
+
+		const newCollectibles = [...collectibles, newEntry];
+		this.update({ collectibles: newCollectibles });
+		return newCollectibles;
+	}
+
+	/**
+	 * Request NFT custom information of a collectible
+	 *
+	 * @param address - Hex address of the collectible contract
+	 * @param tokenId - The NFT identifier
+	 * @returns - Current collectible name and image
+	 */
+	async requestNFTCustomInformation(address: string, tokenId: number): Promise<CollectibleCustomInformation> {
+		if (address in contractMap && contractMap[address].erc721) {
+			const contract = contractMap[address];
+			const api = contract.api;
+			const { name, image } = await this.fetchCollectibleBasicInformation(api, tokenId);
+			return { name, image };
+		} else {
+			return { name: '', image: '' };
+		}
+	}
+
+	/**
+	 * Fetch NFT basic information, name and image url
+	 *
+	 * @param api - API url to fetch custom collectible information
+	 * @param tokenId - The NFT identifier
+	 * @returns - Current collectible name and image
+	 */
+	async fetchCollectibleBasicInformation(api: string, tokenId: number): Promise<CollectibleCustomInformation> {
+		const response = await fetch(this.getCollectibleApi(api, tokenId));
+		const json = await response.json();
+		return json ? { image: json.image_url, name: json.name } : /* istanbul ignore next */ { image: '', name: '' };
```

Comment:
This whole method should fail gracefully. Maybe we could wrap it in a try / catch, and inside the catch, set `name` and `image` to empty strings:

```js
try {
    const response = await fetch(this.getCollectibleApi(api, tokenId));
    const json = await response.json();
    return { image: json.image_url, name: json.name };
} catch(error) {
    return { image: '', name: '' };
}
```

Reply:
should we modify this as well?
https://github.com/MetaMask/gaba/blob/90648e83abd5612fbaad6560df51c663d639f0fa/src/TokenRatesController.ts#L112